File: backend/app/schemas/quiz.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

QuizSubject = Literal["Mathematics", "Science"]
QuizDifficulty = Literal["Easy", "Medium", "Hard"]


class QuizOption(BaseModel):
    id: str
    text: str


class QuizQuestion(BaseModel):
    id: str
    subject: QuizSubject
    topic: str
    difficulty: QuizDifficulty
    prompt: str
    options: list[QuizOption]
    correct_option_id: str
    explanation: str


class QuizGenerateRequest(BaseModel):
    subject: QuizSubject
    topic: str
    difficulty: QuizDifficulty
    question_count: int = Field(ge=1, le=15)


class QuizGenerateResponse(BaseModel):
    questions: list[QuizQuestion]


class QuizSubmitRequest(BaseModel):
    subject: QuizSubject
    topic: str
    difficulty: QuizDifficulty
    total_questions: int = Field(ge=1)
    correct_count: int = Field(ge=0)


class QuizAttemptRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    subject: str
    topic: str
    difficulty: str
    total_questions: int
    correct_count: int
    score_percent: int
    created_at: datetime

File: frontend/src/types/quiz.ts
export type QuizSubject = 'Mathematics' | 'Science'
export type QuizDifficulty = 'Easy' | 'Medium' | 'Hard'

export interface QuizConfig {
  subject: QuizSubject
  topic: string
  difficulty: QuizDifficulty
  questionCount: number
}

export interface QuizOption {
  id: string
  text: string
}

export interface QuizQuestion {
  id: string
  subject: QuizSubject
  topic: string
  difficulty: QuizDifficulty
  prompt: string
  options: QuizOption[]
  correctOptionId: string
  explanation: string
}

export interface QuizAnswer {
  questionId: string
  selectedOptionId: string | null
}

export interface QuizAnswerReview {
  question: QuizQuestion
  selectedOptionId: string | null
  isCorrect: boolean
}

export interface QuizResult {
  totalQuestions: number
  correctCount: number
  scorePercent: number
  answers: QuizAnswerReview[]
}
